cli/cmd/upgrade/upgrade.go

```diff
@@ -298,6 +319,192 @@ func (c *upgradeSpec) UpgradePathBasedOnVersion() error {
+		}
+		delete(releaseValues, "timescaledbExternal")
+	}
+
+	// refactor promscale section in values.yaml
+	var isTracingEnabled bool
+	promscaleValues, _ := releaseValues["promscale"].(map[string]interface{})
+
+	if promscaleValues == nil {
+		// promscale values is nil, construct the spec to
+		// assign db password/uri
+		releaseValues["promscale"] = make(map[string]interface{})
+		promscaleValues, _ := releaseValues["promscale"].(map[string]interface{})
+		promscaleValues["connection"] = make(map[string]interface{})
+		connValues, _ := promscaleValues["connection"].(map[string]interface{})
+		connValues["uri"] = externalDBURI
+		connValues["password"] = tsdbSecretValue
+	} else {
+		for k1, v1 := range promscaleValues {
+			if k1 == "tracing" {
+				traceMap := v1.(map[string]interface{})
+				isTracingEnabled = traceMap["enabled"].(bool)
+
+				promscaleValues["openTelemetry"] = v1
+				delete(promscaleValues, k1)
+			}
+
+			if k1 == "image" {
+				imageName := v1.(string)
+				// In tobs 0.7.0 release we hardcoded
+				// beta release if tracing is enabled, from 0.8.0 tobs
+				// release all default images will include out of the tracing support in
+				// all Promscale images.
+				if imageName == "timescale/promscale:0.7.0-beta.latest" {
+					delete(promscaleValues, k1)
+				}
+			}
+
+			if k1 == "args" {
+				aV, _ := v1.([]interface{})
+				for in, value := range aV {
+					if value == "-otlp-grpc-server-listen-address=:9202" {
+						aV = append(aV[:in], aV[in+1:]...)
+					}
+				}
+				promscaleValues["args"] = aV
+			}
+
+			if k1 == "connection" {
+				connectionValues, _ := v1.(map[string]interface{})
+				connectionValues["uri"] = externalDBURI
+
+				for k2, v2 := range connectionValues {
+					if k2 == "password" {
+						delete(connectionValues, k2)
+						connectionValues["password"] = tsdbSecretValue
+					}
+
+					if k2 == "host" {
+						hostValues, _ := v2.(map[string]interface{})
+						hostValue := hostValues["nameTemplate"]
+						delete(connectionValues, k2)
+						connectionValues["host"] = hostValue
+					}
+				}
+			}
+
+			if k1 == "service" {
+				promService, _ := v1.(map[string]interface{})
+				lbValues := promService["loadBalancer"]
+				lb, _ := lbValues.(map[string]interface{})
+				lbEnabled := lb["enabled"]
+				if lbEnabled == "true" {
+					promService["type"] = "LoadBalancer"
+				} else {
+					promService["type"] = "ClusterIP"
+				}
+				delete(promService, "loadBalancer")
+			}
+		}
+	}
+
+	if isTracingEnabled {
+		otelCol := otel.OtelCol{
+			ReleaseName: root.HelmReleaseName,
+			Namespace:   root.Namespace,
+			K8sClient:   c.k8sClient,
+			HelmClient:  helmClient,
+		}
+
+		// Kube-Prometheus CRDs
+		err = c.createCRDS(kubePrometheusCRDs)
+		if err != nil {
+			return err
+		}
+
+		err = otelCol.UpgradeCertManager()
+		if err != nil {
+			return err
+		}
+
+		// OpenTelemetry CRDs
+		err = c.createCRDS(openTelemetryCRDs)
+		if err != nil {
+			return err
+		}
```

Comment:
nit: Can we move this to be before upgrading cert-manager and after creation of kube-prometheus CRDs?